glide-core/redis-rs/redis/src/cluster_async/pipeline_routing.rs

```diff
@@ -0,0 +1,519 @@
+) -> (
+    Vec<oneshot::Receiver<RedisResult<Response>>>,
+    Vec<PendingRequest<C>>,
+    Vec<(String, Vec<(usize, Option<usize>)>)>,
+)
+where
+    C: Clone + ConnectionLike + Connect + Send + Sync + 'static,
+{
+    let mut receivers = Vec::new();
+    let mut pending_requests = Vec::new();
+    let mut addresses_and_indices = Vec::new();
+
+    for (address, context) in pipelines_by_connection {
+        // Create a channel to receive the pipeline execution results
+        let (sender, receiver) = oneshot::channel();
+        // Add the receiver to the list of receivers
+        receivers.push(receiver);
+        pending_requests.push(PendingRequest {
+            retry: 0,
+            sender,
+            info: RequestInfo {
+                cmd: CmdArg::Pipeline {
+                    count: context.pipeline.len(),
+                    pipeline: context.pipeline.into(),
+                    offset: 0,
+                    route: InternalSingleNodeRouting::Connection {
+                        address: address.clone(),
+                        conn: async { context.connection }.boxed().shared(),
+                    },
+                    // mark it as a sub-pipeline mode
+                    sub_pipeline: true,
+                },
+            },
+        });
+        // Record the node address and its associated command indices for result mapping
+        addresses_and_indices.push((address, context.command_indices));
+    }
+
+    (receivers, pending_requests, addresses_and_indices)
+}
+
+/// Adds the result of a pipeline command to the `pipeline_responses` collection.
+///
+/// This function updates the `pipeline_responses` vector at the given `index` and optionally at the
+/// `inner_index` if provided. If `inner_index` is `Some`, it ensures that the vector at that index is large enough
+/// to hold the value and address at the specified position, resizing it if necessary. If `inner_index` is `None`,
+/// the value and address are simply appended to the vector.
+///
+/// # Parameters
+/// - `pipeline_responses`: A mutable reference to a vector of vectors that stores the results of pipeline commands.
+/// - `index`: The index in `pipeline_responses` where the result should be stored.
+/// - `inner_index`: An optional index within the vector at `index`, used to store the result at a specific position.
+/// - `value`: The result value to store.
+/// - `address`: The address associated with the result.
+pub fn add_pipeline_result(
+    pipeline_responses: &mut PipelineResponses,
+    index: usize,
+    inner_index: Option<usize>,
+    value: Value,
+    address: String,
+) {
+    match inner_index {
+        Some(inner_index) => {
+            // Ensure the vector at the given index is large enough to hold the value and address at the specified position
+            if pipeline_responses[index].len() <= inner_index {
+                pipeline_responses[index].resize(inner_index + 1, (Value::Nil, "".to_string()));
```

Comment:
add todo - change the pipeline_responses to hold in [index] a vector already sized with the expected responses len

Reply:
on the next PR change Value::Nil to error